synapse/storage/databases/main/events_worker.py

```diff
@@ -754,7 +754,31 @@ def _invalidate_local_get_event_cache(self, event_id: str) -> None:
     async def _get_events_from_cache(
         self, events: Iterable[str], update_metrics: bool = True
     ) -> Dict[str, EventCacheEntry]:
-        """Fetch events from the caches.
+        """Fetch events from the caches, both in memory and any external.
+
+        May return rejected events.
+
+        Args:
+            events: list of event_ids to fetch
+            update_metrics: Whether to update the cache hit ratio metrics
+        """
+        event_map = self._get_events_from_local_cache(events, update_metrics=update_metrics)
+
+        missing_event_ids = {e for e in events if e not in event_map}
+
+        for event_id in missing_event_ids:
+            ret = await self._get_event_cache.get(
+                (event_id,), None, update_metrics=update_metrics
+            )
```

Comment:
this looks like redundant work for cache implementations that *don't* have an external cache.